File: leetcode/java/0011. Container With Most Water.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(1)

class Solution {
    public int maxArea(int[] height) {
        int left = 0;
        int right = height.length - 1;

        int maxArea = 0;

        while (right > left) {
            int width = right - left;
            int tempArea = width * Integer.min(height[left], height[right]);

            maxArea = Integer.max(tempArea, maxArea);

            if (height[left] > height[right]) {
                right -= 1;
            } else {
                left += 1;
            }
        }
        return maxArea;
    }
}

File: leetcode/python/0011. Container With Most Water.py
# Time complexity: O(N)
# Space complexity: O(1)

class Solution(object):
    def maxArea(self, height):
        maxArea = 0
        width = len(height) - 1

        left = 0
        right = len(height) - 1

        while right > left:
            tempArea = min(height[left], height[right]) * width
            if height[left] > height[right]:
                right -= 1
            else:
                left += 1
            maxArea = max(tempArea, maxArea)
            width -= 1
        return maxArea